File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        int firstCritical = -1;
        int lastCritical = -1;

        int minDistance = Integer.MAX_VALUE;

        int index = 1;

        ListNode prev = head;
        ListNode curr = head.next;

        while (curr != null && curr.next != null) {

            ListNode next = curr.next;

            // check if current nde is a critical point
            if ((curr.val > prev.val && curr.val > next.val) || 
                (curr.val < prev.val && curr.val < next.val)) {

                // first critical point
                if (firstCritical == -1) {
                    firstCritical = index;
                } else {
                    // distance from previous critical point
                    minDistance = Math.min(minDistance, index - lastCritical);
                }

                lastCritical = index;
            }

            prev = curr;
            curr = next;
            index++;
        }

        // fewer than two critical points
        if (firstCritical == lastCritical)
            return new int[]{-1, -1};
        
        int maxDistance = lastCritical - firstCritical;

        return new int[]{minDistance, maxDistance};
    }
}

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number[]}
 */
var nodesBetweenCriticalPoints = function(head) {
    let firstCritical = -1;
    let lastCritical = -1;

    let minDistance = Infinity;

    let index = 1;

    let prev = head;
    let curr = head.next;

    while (curr !== null && curr.next !== null) {
        const next = curr.next;

        // check if current node is a critical point
        if ((curr.val > prev.val && curr.val > next.val) || 
            (curr.val < prev.val && curr.val < next.val)) {

            // first critical point
            if (firstCritical === -1) {
                firstCritical = index;
            } else {
                // distance from previous critical point
                minDistance = Math.min(minDistance, index - lastCritical);
            }

            lastCritical = index;
        }

        prev = curr;
        curr = next;
        index++;
    }

    // fewer than two critical points
    if (firstCritical === lastCritical) {
        return [-1, -1];
    }
    
    const maxDistance = lastCritical - firstCritical;

    return [minDistance, maxDistance];
};
